portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -43,59 +47,74 @@
     Query query = new Query(COMMENT).addSort(TIMESTAMP, SortDirection.DESCENDING);
 
     DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
     PreparedQuery results = datastore.prepare(query);
+    
+    //String languageCode = request.getParameter("languageCode");
+    String languageCode = request.getParameter("languageCode");
 
     List<Comment> comments = new ArrayList<>();
     for (Entity entity : results.asIterable()) {
       String username = (String) entity.getProperty(USERNAME);
       String commentBody = (String) entity.getProperty(COMMENT_BODY); 
       String timestamp = (String) entity.getProperty(TIMESTAMP);
 
-      Comment comment = new Comment(username, commentBody, timestamp);
+      // Do the translation for comment body.
+      Translation commentBodyTranslation =
+        translate.translate(commentBody, Translate.TranslateOption.targetLanguage(languageCode));
+      String commentBodyTranslatedText = commentBodyTranslation.getTranslatedText();
+
+      // Do the translation for timestamp.
+      Translation timestampTranslation =
+        translate.translate(timestamp, Translate.TranslateOption.targetLanguage(languageCode));
+      String timestampTranslatedText = timestampTranslation.getTranslatedText();
+
+      Comment comment = new Comment(username, commentBodyTranslatedText, timestampTranslatedText);
       comments.add(comment);
     }
 
     Gson gson = new Gson();
 
-    response.setContentType("application/json;");
+    response.setContentType("application/json; charset=UTF-8");
+    response.setCharacterEncoding("UTF-8");
     response.getWriter().println(gson.toJson(comments));
   }
 
   @Override
   public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
     // Get the input from the form
-    String username = getParameter(request, "user-name").trim();
-    String commentBody = getParameter(request, "text-input").trim();  
+    String username = request.getParameter("user-name").trim();
+    String commentBody = request.getParameter("text-input").trim();  
     
     Date date = new Date();
     SimpleDateFormat dateFormat = new SimpleDateFormat("EEE, dd/MM/yyyy hh:mm z");  
     String timestamp = dateFormat.format(date).toString();
     
+    // Entity for the comment.
     Entity commentEntity = new Entity(COMMENT);
     commentEntity.setProperty(USERNAME, username);
     commentEntity.setProperty(COMMENT_BODY, commentBody);
     commentEntity.setProperty(TIMESTAMP, timestamp);
 
+    // Entity for the current language code chosen.
+    Entity languageEntity = new Entity(LANGUAGE);
+    languageEntity.setProperty(LANGUAGE_CODE, request.getParameter("languageCode"));
+
     DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
     datastore.put(commentEntity);
+    datastore.put(languageEntity);
     
     // Redirect back to the HTML page.
     response.sendRedirect("/index.html#comment-section");
```

Comment:
If you want to persist language, the returned url should contain language as a url parameter. e.g. `/index.html#comment-section?lang=CN` and client side JS needs to parse the language parameter in order to set the language inside the request to the server.